lib/token.js

```diff
@@ -243,7 +241,80 @@ function addFragmentListener(sdk, windowEl, timeout) {
   return deferred.promise.timeout(timeout || 120000, new AuthSdkError('OAuth flow timed out'));
 }
 
-function handleOAuthResponse(sdk, oauthParams, res) {
+function getOAuthUrls(sdk, oauthParams, options) {
+  options = options || {};
+
+  // Get user-supplied arguments
+  var authorizeUrl = util.removeTrailingSlash(options.authorizeUrl) || sdk.options.authorizeUrl;
+  var issuer = util.removeTrailingSlash(options.issuer) || sdk.options.issuer;
+  var userinfoUrl = util.removeTrailingSlash(options.userinfoUrl) || sdk.options.userinfoUrl;
+
+  // If an issuer exists, but it's not a url
+  if (issuer && !(/^https?:/.test(issuer))) {
+    // Make it a url
+    issuer = sdk.options.url + '/oauth2/' + issuer;
+  }
+
+  // If an authorizeUrl is supplied without an issuer, and an id_token is requested
+  if (authorizeUrl &&
+      !issuer &&
+      oauthParams.responseType.indexOf('id_token') !== -1) {
+    // The issuer is ambiguous, so we won't be able to validate the id_token jwt
+    throw new AuthSdkError('Cannot request idToken with an authorizeUrl without an issuer');
+  }
+
+  // If a token is requested without an issuer
+  if (!issuer && oauthParams.responseType.indexOf('token') !== -1) {
+    // If an authorizeUrl is supplied without a userinfoUrl
+    if (authorizeUrl && !userinfoUrl) {
+      // The userinfoUrl is ambiguous, so we won't be able to call getUserInfo
+      throw new AuthSdkError('Cannot request accessToken with an authorizeUrl without an issuer or userinfoUrl');
+    }
+
+    // If a userinfoUrl is supplied without a authorizeUrl
+    if (userinfoUrl && !authorizeUrl) {
+      // The authorizeUrl is ambiguous, so we won't be able to call the authorize endpoint
+      throw new AuthSdkError('Cannot request token with an userinfoUrl without an issuer or authorizeUrl');
+    }
+  }
+
+  var isSharedResourceServer = new RegExp('^https?://.*?/oauth2/.+');
+
+  // If issuer is supplied, but not authorizeUrl
+  if (issuer && !authorizeUrl) {
+    // Build authorizeUrl until we create it dynamically
+    if (isSharedResourceServer.test(issuer)) {
+      authorizeUrl = issuer + '/v1/authorize';
+    } else {
+      authorizeUrl = issuer + '/oauth2/v1/authorize';
```

Comment:
In this case since we're doing this multiple times, it would be nice to capture this once a little earlier around the line with sharedResourceServerRe and set a "baseAuthServerUrl" - if we do that, all this logic is basically:
```javascript
if (!userinfoUrl) {
  userinfoUrl = baseAuthServerUrl + '/v1/userinfo';
}
```

Also, I notice down below that we're repeating this for the default logic. Seems like the general idea should be:

1. Get urls from either function options or client options
2. Throw errors if they're in a bad state
3. Set baseAuthServerUrl
4. Fill in any values that aren't set - the baseUrl will take care of the default values